### DISCUSSION TITLE
Is it possible to use OTA for the Pi Pico using PlatformIO

### DISCUSSION THREAD

**Discussion by mblees** [earlephilhower/arduino-pico#2556] [General]
Hello there,
I have found that its possible to use OTA for the PiPico when using the ArduinoIDE. However i am using PlatformIO and have a pretty sizable project in there that i wouldn't want to migrate.

I know for the ESP32 you can simple change the upload_protocol in the platformio.ini to espota and you can use OTA to flash the controller. Is there the same functionality already established for the Pi Pico family?

specifics:
controller: wiznet_5100s_evb_pico
network: ethernet

This is my first try at building the ota functionality for the controller
```
void init_ota(void){
    ArduinoOTA.setHostname(ota_hostname);
    ArduinoOTA.setPassword(ota_password);
    ArduinoOTA.begin(Ethernet.localIP());
}

void handle_ota(void){
    ArduinoOTA.handle();
}
```

And this is my current platformio.ini
```
[env:wiznet_5100s_evb_pico]
platform = raspberrypi
board = wiznet_5100s_evb_pico
framework = arduino
```

If you have any informations, it would be highly appreciated.
Poirot

**Comment by maxgerhardt** [comment]
The platform has an example exactly for that. It uses the same `espota` upload protocol.

https://github.com/maxgerhardt/platform-raspberrypi/tree/develop/examples/arduino-ota

We even support fancy signed OTA binaries: https://github.com/maxgerhardt/platform-raspberrypi/tree/develop/examples/arduino-signed-ota

**Reply by mblees** [reply]
Thanks for your answer but i am still confused. The example you linked is using WiFi of the PicoW, while I am using the Ethernet Shield attached to the W5100S-EVB-Pico.
Are Ethernet and Wifi exchangable in this case?

I've tried simply swapping all the WiFi stuff in the example with Ethernet functions but to no success since i get stuck in the ArduinoOTA.begin() function.

```
Mac Adress: 222, 4, 37, 82, 161, 49, 
Initialize Ethernet with DHCP:       
My IP address: 192.168.xxx.xxx
Arrived in begin(0)
Arrived in begin(1)
Arrived in begin(1.1)
Arrived in begin(1.2)
```


update: Using this code https://github.com/mblees/W5100S-EVB-Pico-OTE I manged to make the controller show up in the ArduinoIDE. Im not yet able to flash to it using OTA.

![Screenshot 2024-10-28 113158](https://github.com/user-attachments/assets/f9e4b7db-f1bf-4c61-b23e-2ef6df38455f)

**Reply by maxgerhardt** [reply]
`espota.py` doesn't know that it's talking to a Pico W over WiFI or a W5100S over Ethernet -- as long as the chip responds the same, it will be fine. I have a Wiznet W5500-EVB-Pico (`wiznet_5500_evb_pico`) that I will now test with. This is not exactly wiznet_5100s_evb_pico but close I guess.

**Reply by maxgerhardt** [reply]
I have just personally verified that OTA on the W5500-EVB-Pico via Ethernet works perfectly fine. See

https://github.com/maxgerhardt/pio-w5500-evb-pico-ota

Try to **not** use the standard `Ethernet.h` library, use the LWIP library already built into the core. I'm using
```cpp
#include <W5500lwIP.h>
#define WIZNET_RSTn 20
#define WIZNET_CS   17
#define WIZNET_INT  21

Wiznet5500lwIP eth(WIZNET_CS /* chip select */, SPI, WIZNET_INT /* interrupt */);

void setup() {
  SPI.setRX(16);
  SPI.setCS(17);
  SPI.setSCK(18);
  SPI.setTX(19);
  //..
}
```
here but you'll have to change that for your board to use `#include <W5100lwIP.h>` accordingly, with the right pins.

The sketch also makes the board show up in the Arduino IDE.

![grafik](https://github.com/user-attachments/assets/63873301-ab2a-45e1-bdbd-1d47936b03e3)

**Reply by maxgerhardt** [reply]
Ah, I've just made it easier for you and added support for the WIZnet 5100S EVB as well. Assuming that the 5100S is compatible with the 5100 code? No idea.

Please download the repo above and first upload + monitor the `wiznet_5100s_evb_pico_via_usb` environment to see if Ethernet works, then update the IP in `wiznet_5100s_evb_pico_via_ota`  and try to upload again.

**Reply by mblees** [reply]
You are seriously amazing, Thank you very much!

Worked first try!

**Reply by mblees** [reply]
Following up with this project I have ran into another issue. Since im now using the W5100lwIP instead of the modified Wiznet Ethernet that I had been using before: 
I can't seam to use the PubSubClient anymore since the Wiznet5100lwIP class does not have the required client attribute that the other Ethernet libs provide.

My old MQTT implementation worked like this: 
![image](https://github.com/user-attachments/assets/edfce65b-c7fe-47a2-af54-d4ae5f13d50c)

However PubSubClient requires the client attribute which the W5100lwIP lib doesnt provide:
![image](https://github.com/user-attachments/assets/dd9678da-07c0-4217-bd90-19a8aaf378e3)

**Reply by maxgerhardt** [reply]
This may seem counter-intuitive, but you actually need to create a `WiFiClient` instead of an `EthernetClient`. The `WiFiClient` is just an interface to the lwIP stack. The lwIP stack in your case then uses the W5100 as its transport, or an actual WiFi chip in case of a Pico W. But the interface stays the same. The `WiFiClient` is provided by `#include <WiFi.h>`, which is already auto-included by `ArduinoOTA.h`

**Reply by maxgerhardt** [reply]
I pushed the example code in https://github.com/maxgerhardt/pio-w5500-evb-pico-ota/tree/mqtt and it works fine on my board with a public MQTT test server by https://www.shiftr.io/try?lang=terminal.

```
--- Quit: Ctrl+C | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H
Booting
Waiting for Ethernet link..
Waiting for Ethernet link..
Waiting for DHCP address..
Ethernet Ready!
IP address: 192.168.0.208
Attempting MQTT connection...connected
Published 'hello world' to topic "/hello" OK!
Subscribing to topic "/hello" OK!
Message arrived [/hello] test message from desktop
```

When triggered as
```
curl -X POST "https://public:public@public.cloud.shiftr.io/broker/hello" -d "test message from desktop"
```

**Reply by mblees** [reply]
Awesome thank you very much!
